src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    internal class DependeciesGroupsForEachTFMRule : IPackageRule
+    {
+        private static readonly NuGetFramework Net00 = NuGetFramework.Parse("net00");
+
+        public string MessageFormat => AnalysisResources.DependenciesGroupsForEachTFMHasNoExactMatch;
+
+        public string CompatMatchFoundWarningMessageFormat => AnalysisResources.DependenciesGroupsForEachTFMHasCompatMatch;
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var files = builder.GetFiles().
+                Where(t => PathUtility.GetPathWithDirectorySeparator(t).Count(m => m == Path.DirectorySeparatorChar) > 1);
+            var packageNuspec = ExtractTFMsFromNuspec(builder.GetNuspec());
+            var (compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec) = CatagoriseTFMs(files, packageNuspec);
+            return GenerateWarnings(compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec);
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(HashSet<NuGetFramework> compatNotExactMatches, HashSet<NuGetFramework> noExactMatchesFromFile, HashSet<NuGetFramework> noExactMatchesFromNuspec)
```

Comment:
Now that this method is short, I better understand what's going on, and there's a lot of string allocations going on. One of the most common causes for app performance problems in managed memory runtimes is garbage collection, and garbage collection happens more frequently with object allocation. Therefore, performance optimisation is achieved by minimising the number of objects created. For strings, that means using `StringBuilder`.

Consider how the NU5130 warning is created. In `GenerateWarningString` a `StringBuilder` is created, which is great. However, the variable `ending` is assigned a temporary string created by concatenating two strings. Then when `compatNotExactMatches.Count > 1`, `String.Join` is called where the first parameter is the result of another concatenation. The second parameter is a LINQ query that gets converted to a temporary array using `ToArray`, which internally first creates a collection to determine the array size, since arrays are not resizable in .NET. The result of the `String.Join` is concatenated with another string, before being passed to `StringBuilder.Append`. `StringBuilder.ToString()` is called to return from `GenerateWarningString`, but used again in another `String.Join()` before the final result is passed to `PackagingLogMessage.CreateWarning()`. So, I can count about 6 temporary strings allocated in the creation of this message, not including non-string objects created by the `ToArray()` call. The generation of the warning for NU5128 is similar, but has an addition duplication of the least efficient block.

`String.Join` is a convenient API to create a string, but returns a new string, so not optimal for using with a `StringBuilder`. It's better to use a loop in this case.

Additionally, the conditions in this method (`if (noExactMatchesFromFile.Count != 0 || noExactMatchesFromNuspec.Count != 0)` and `if (compatNotExactMatches.Count != 0`). Mean that the strings created by `GenerateWarningString` won't be used. Fortunately that method has effectively the same conditions so the worst of the string allocations won't happen, but if you merge the two methods in here, you can avoid creating `StringBuilder` and any temporary strings (like `beginning` and `end`) unless you know for a fact it's going to be used.

Looking at `GenerateWarningString`, it appears the generation of the two strings are completely independent, just as the creation of the `NU5128` and `NU5130` warnings are below, so I suggest avoid mixing them so it's easier to understand.